src/inspection/inspectionLinks.ts

```
@@ -0,0 +1,202 @@
+  let inspectionWeekLinks = inspection.weekInspectionMap || {}
+
+  if (isEmpty(inspectionWeekLinks)) {
+    inspectionWeekLinks = await getLinkedInspectionMapForPeriod(inspection, false)
+  }
+
+  let inspectionLinks: LinkedInspectionForWeek[] = []
+
+  for (let [startDate, inspectionIdOrInspection] of Object.entries(inspectionWeekLinks)) {
+    let inspectionLink = new LinkedInspectionForWeek()
+
+    let linkedInspection: Inspection =
+      typeof inspectionIdOrInspection === 'string'
+        ? await inspectionRepo.findOneOrFail(inspectionIdOrInspection)
+        : inspectionIdOrInspection
+
+    inspectionLink.id = v4()
+    inspectionLink.startOfWeek = startDate
+    inspectionLink.inspection = linkedInspection
+
+    inspectionLinks.push(inspectionLink)
+  }
+
+  return inspectionLinks
+}
+
+export async function getLinkedInspectionMapForPeriod(
+  inspection: Inspection,
+  useIds = false // Set only the inspection ID on the mapping instead of the inspection entity.
+): Promise<WeekInspectionMap> {
+  let inspectionType = getLinkedInspectionType(inspection)
+  let linkedInspectionStackIdentifier = getLinkedInspectionStackIdentifier(inspection)
+
+  // If inspectionType is falsy, the stack identifier is also falsy, but we need to keep Typescript happy.
+  if (!inspectionType || !linkedInspectionStackIdentifier) {
+    return {}
+  }
+
+  let inspectionInterval = {
+    start: parseISO(inspection.inspectionStartDate),
+    end: parseISO(inspection.inspectionEndDate),
+  }
+
+  let inspectionWeeks = eachWeekOfInterval(inspectionInterval, { weekStartsOn: 1 })
+  let inspectionWeekMap = {}
+
+  for (let week of inspectionWeeks) {
+    let weekEndDate = addDays(week, 7)
+
+    let productionInspection = await getProductionInspectionForPeriod({
+      stackIdentifier: linkedInspectionStackIdentifier,
+      inspectionType,
+      startDate: week,
+      endDate: weekEndDate,
+    })
+
+    if (!productionInspection) {
```

Hmm, this is something that should never happen right? And if it happens, then we have some issue somewhere. I think we could safely throw an error here to indicate that something is wrong

---

This will happen if the user attempts to create a post-inspection for a period where there are no pre-inspections in effect. In this case the inspection validation already notifies the user about this and they can either change the inspection period or try again when there are pre-inspections available. I don't think an additional error here is necessary, as we already have validation and user guidance for this case.

---

Yeah this is why this shouldn't never happen. But if a developer makes a mistake, this can happen (this is why I would throw an error). It's something I call as sanity check

---

My point was that it doesn't hurt throwing error there. Also the error msg indicates to the reader that it shouldn't never happen